File: Niveau 3/Niveau 3.9/milieux.py
#-------------------------------------------------------------------------------
# Name:        Les bons milieux
# Purpose:     France-IOI niveau 3.9
#
# Author:      Sébastien
#
# Created:     10/10/2022
#-------------------------------------------------------------------------------

import sys

def main():
    taille = 100
    nbPoints = int(sys.stdin.readline())
    plan = [[False]*(taille+1) for _ in range(taille+1)]
    liste = []
    for _ in range(nbPoints):
        x, y = map(int, sys.stdin.readline().split())
        liste.append((x, y))
        plan[x][y] = True

    nbMilieux = 0
    for p in liste:
        for q in liste:
            if q != p:
                # On calcule les coordonnées du point S symétrique de Q par rapport à P
                xs = 2*p[0] - q[0]
                ys = 2*p[1] - q[1]
                # Si S est dans le plan, alors P est un milieu (milieu de [QS])
                if xs >= 0 and xs <= taille and ys >= 0 and ys <= taille and plan[xs][ys]:
                    nbMilieux += 1
                    break

    print(nbMilieux)

if __name__ == '__main__':
    main()

"""
7
0 4
2 4
4 4
4 0
1 2
2 2
3 2
---
3
50 60
52 60
52 58
"""

File: Niveau 3/Niveau 3.9/milieux.hs
--------------------------------------------------------------------------------
-- Name:        Les bons milieux
-- Purpose:     France-IOI niveau 3.9
--
-- Author:      Sébastien
--
-- Created:     11/10/2022
--------------------------------------------------------------------------------

import Control.Monad
import Data.Array

type Point = (Int,Int)
taille = 100

main :: IO()
main = do
    nbPoints <- readLn :: IO Int
    liste <- replicateM nbPoints lirePoint
    let plan = listArray ((0,0), (taille,taille)) (repeat False)
    let plan' = plan // [(point, True) | point <- liste]
    let nbMilieux = foldr (\p n -> n + (estMilieu p liste plan')) 0 liste
    print nbMilieux
    
lirePoint :: IO Point
lirePoint = do (x:y:_) <- fmap (map read . words) getLine :: IO [Int]
               return (x,y)

estMilieu :: Point -> [Point] -> Array Point Bool -> Int -- Retourne 1 si le point est milieu, 0 sinon
estMilieu _ [] _ = 0
estMilieu p (q:qs) plan
    | p == q = estMilieu p qs plan
    | otherwise = let xs = 2 * fst p - fst q
                      ys = 2 * snd p - snd q
                  in if xs >= 0 && xs <= taille && ys >= 0 && ys <= taille && plan ! (xs,ys)
                     then 1
                     else estMilieu p qs plan                  
    